src/Esprima/Utils/JavascriptTextWriter.cs

```diff
@@ -0,0 +1,375 @@
+
+public delegate object? NodePropertyValueAccessor(Node node);
+
+public delegate ref readonly NodeList<T> NodePropertyListValueAccessor<T>(Node node) where T : Node?;
+
+/// <summary>
+/// Base Javascript text writer (code formatter) which uses the most compact possible (i.e. minimal) format.
+/// </summary>
+public partial class JavascriptTextWriter
+{
+    public record class Options
+    {
+        public static readonly Options Default = new();
+    }
+
+    public delegate JavascriptTextWriter Factory(TextWriter writer, Options options);
+
+    private readonly TextWriter _writer;
+
+    public JavascriptTextWriter(TextWriter writer, Options options)
+    {
+        _writer = writer ?? throw new ArgumentNullException(nameof(writer));
+
+        if (options is null)
+        {
+            throw new ArgumentNullException(nameof(options));
+        }
+
+        LastTokenType = TokenType.EOF;
+        WhiteSpaceWrittenSinceLastToken = true;
+    }
+
+    protected TokenType LastTokenType { [MethodImpl(MethodImplOptions.AggressiveInlining)] get; [MethodImpl(MethodImplOptions.AggressiveInlining)] private set; }
+    protected TokenFlags LastTokenFlags { [MethodImpl(MethodImplOptions.AggressiveInlining)] get; [MethodImpl(MethodImplOptions.AggressiveInlining)] private set; }
+    protected bool WhiteSpaceWrittenSinceLastToken { [MethodImpl(MethodImplOptions.AggressiveInlining)] get; [MethodImpl(MethodImplOptions.AggressiveInlining)] private set; }
+
+    #region White-space
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    protected void WriteSpace()
+    {
+        WriteWhiteSpace(" ");
+    }
+
+    protected void WriteLine()
+    {
+        _writer.WriteLine();
+        WhiteSpaceWrittenSinceLastToken = true;
+    }
+
+    protected void WriteWhiteSpace(string value)
+    {
+        _writer.Write(value);
+        WhiteSpaceWrittenSinceLastToken = true;
+    }
+
+    #endregion
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    protected void ForceRecommendedSpace()
+    {
+        LastTokenFlags |= TokenFlags.TrailingSpaceRecommended;
+    }
+
+    public virtual void WriteEpsilon(TokenFlags flags, in WriteContext context) { }
+
+    #region Identifiers
+
+    protected virtual void StartIdentifier(string value, TokenFlags flags, in WriteContext context)
+    {
+        switch (LastTokenType)
+        {
+            case TokenType.BigIntLiteral:
+            case TokenType.BooleanLiteral:
+            case TokenType.Identifier:
+            case TokenType.Keyword:
+            case TokenType.NullLiteral:
+            case TokenType.NumericLiteral:
+            case TokenType.RegularExpression:
+                WriteSpace();
+                break;
+            case TokenType.EOF:
+            case TokenType.Punctuator:
+            case TokenType.StringLiteral:
+            case TokenType.Template:
+                break;
+            default:
+                throw new InvalidOperationException();
```

Comment:
should prefer `EsprimaExceptionHelper` for consistency, this is argument out of range I guess?

Reply:
See https://github.com/sebastienros/esprima-dotnet/pull/291#discussion_r924382111